### PR TITLE
Async Hamburger Menu Implementation + Caching Optimizations

### PR DESCRIPTION
**Pull request by Ridhwana**
Closed [this PR ](https://github.com/forem/forem/pull/18760/files) in favor of opening thsi one that deals with the async hamburger and caching optimizations.

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR sends HTML over an endpoint /async_info/navigation_links and renders it dynamically when we click the hamburger menu. 

It also implements both Rails caching for the sql query as well as edge caching. After some testing from Mac and I, the browser caching looks like its working as well.

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/18639

## QA Instructions, Screenshots, Recordings

1. Go to `http://localhost:3000/`
2. Update your screen to be a mobile viewport
3.  Click on the hamburger menu 

Expected behaviour: Make sure that it makes a request to `/async_info/navigation_links` and returns the HTML that gets rendered in the sidebar. 

Caching notes: 

The Fastly caching can only be tested in production. 
We had temporarily deployed to the benhalpern community and checked the caching attributes:
Screenshot 2022-12-13 at 16 48 49

The browser caching relies on the Fastly caching headers being set and so that can also only be tested in production. We had temporarily deployed to the benhalpern community and checked the caching attributes:
Screenshot 2022-12-13 at 16 51 43

The Rails caching can be tested by looking at the sql queries being logged and noting in the rails logs when the cache has been hit.

### UI accessibility concerns?

None

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/controllers/async_info_controller.rb
Comment (Ridhwana) [inline app/controllers/async_info_controller.rb:67]: You'll notice that I took out the query for `@navigation_links`. It was a mistake to add it as the query is actually being done in the view for caching purposes.

File: app/controllers/async_info_controller.rb
Comment (Ridhwana) [inline app/controllers/async_info_controller.rb:4]: I'd like to test the edge caching on the benhalpern community.

File: app/views/async_info/navigation_links.html.erb
Comment (Ridhwana) [inline app/views/async_info/navigation_links.html.erb:1]: Added some Fragment caching for the hamburger menu. Its useful to note that this is also being done in https://github.com/forem/forem/blob/b03839ba9f6ace4f9ead3a656e69c0fae30c73e5/app/views/articles/_sidebar.html.erb#L3 for the main_nav etc.

**Review by Ridhwana — COMMENTED** [review]
This is still in draft but would love some sanity checks from @fdocr and @maestromac please.

File: app/views/layouts/_main_nav.html.erb
Comment (maestromac) [reply app/views/layouts/_main_nav.html.erb:2]: This would be Russian doll caching which should be an optimization but in this cause I think it would be a negligible effect. We can overlook it for now.

File: app/views/layouts/_main_nav.html.erb
Comment (Ridhwana) [inline app/views/layouts/_main_nav.html.erb:7]: **Another alternative:**

Instead of having two separate caches here, one for the "default nav links" and the other for the "other nav links" we could just have one "navigation_links" cache like how I had it previously:


````
@navigation_links = Rails.cache.fetch("navigation_links", expires_in: 5.minutes) do
  {
    default_nav_links: NavigationLink.default_section.ordered.ids,
  }
end

@navigation_links = Rails.cache.fetch("navigation_links", expires_in: 5.minutes) do
  {
    other_navigation_links: NavigationLink.other_section.ordered.ids,
  }
end
```

And then in the app/controllers/admin/navigation_links_controller.rb above, when we delete the cache, we wouldn't check what type of link has been CRUD'ed, and update that specific cache, but we'd just delete the whole `navigation_links` cache with  `Rails.cache.delete("navigation_links")`.

On the surface, the way that I have it currently seem optimal because we wont be purging *all* the navigation links but we'd be purging it by the appropriate section. However, maybe that's introducing an extra dependency or not optimal in terms of spinning up two caches. I'd love more thoughts on this please?

Comment (Ridhwana) [reply app/views/layouts/_main_nav.html.erb:7]: cc @maestromac

Comment (maestromac) [reply app/views/layouts/_main_nav.html.erb:7]: Didn't mean to confuse you here which the example I sent you prior but yes combining them is the most optimal way.

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [inline app/javascript/packs/base.jsx:72]: We no longer have a "more" button in the navigation, we had removed this functionality a while back but this code was left behind. There is no longer a js-nav-more-trigger in the code.

File: app/javascript/topNavigation/utilities.js
Comment (Ridhwana) [inline app/javascript/topNavigation/utilities.js:125]: Removing old code - i.e. "More" menu in the navigation.

File: spec/system/ahoy/tracking_create_account_spec.rb
Comment (Ridhwana) [inline spec/system/ahoy/tracking_create_account_spec.rb:16]: I this belongs in the Cypress tests, and we can move it away from Capybara, for now I updated it here but when I take a look at best practices we can perhaps look into where it actually belongs

File: spec/system/ahoy/tracking_create_account_spec.rb
Comment (Ridhwana) [inline spec/system/ahoy/tracking_create_account_spec.rb:44]: This is Rubocop asking for alphabetical attributes.

File: app/controllers/admin/navigation_links_controller.rb
Comment (Ridhwana) [inline app/controllers/admin/navigation_links_controller.rb:17]: We want to delete the cache every time theres a CRUD action on navigation_links

File: app/controllers/async_info_controller.rb
Comment (Ridhwana) [inline app/controllers/async_info_controller.rb:4]: We add some cache control headers to enable Fastly (edge) caching

File: app/views/layouts/_main_nav.html.erb
Comment (Ridhwana) [inline app/views/layouts/_main_nav.html.erb:1]: We added some caching here on the sql queries as well.

File: app/views/layouts/_main_nav.html.erb
Comment (Ridhwana) [inline app/views/layouts/_main_nav.html.erb:10]: Collection caching on this partial

File: app/controllers/admin/navigation_links_controller.rb
Comment (maestromac) [reply app/controllers/admin/navigation_links_controller.rb:17]: non-blocking suggestion: seems like it is part of `after_action` and I see there's something similar there already.

File: app/views/layouts/_main_nav.html.erb
Comment (maestromac) [reply app/views/layouts/_main_nav.html.erb:2]: After looking at it again, I want to point out that these two cache are not directly related to each other and may not clear together when you expect them to

**Review by fdocr — APPROVED** [review]
LGTM pending resolution of [this conversation](https://github.com/forem/forem/pull/18818#discussion_r1043543424) regarding the caching strategy on those partials

**Review by benhalpern — CHANGES_REQUESTED** [review]

File: app/controllers/async_info_controller.rb
Comment (Ridhwana) [reply app/controllers/async_info_controller.rb:4]: Thanks for catching this Ben! I'd also love to discuss the other possibilities, once this is merged, I'll start a thread in #eng.

**Review by benhalpern — APPROVED** [review]
LGTM!

**Pull request merged**
